### PR TITLE
CDAP Programs and Components to use changed metrics collection API

### PR DESCRIPTION
**Pull request by abaranau**
Follow up for https://github.com/caskdata/cdap/pull/915. In this PR all client code that collects metrics is adjusted to use changed collect API (to see the API change, refer to https://github.com/caskdata/cdap/pull/915).

Next PRs will contain metrics transfer changes (since metrics record is changed to carry the context as tags), metrics processing (we map raw metric values to the aggregations we want to compute on server side) and test framework adjustments.

JIRAs to create:
* report service handler response metrics per status
* change TxMetricsCollector interface in Tephra

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 1]
```diff
@@ -57,11 +57,12 @@
   private final DatasetInstantiator dsInstantiator;
   private final DiscoveryServiceClient discoveryServiceClient;
 
+
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: remove newlines?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 2]
```diff
@@ -179,4 +170,13 @@ protected void closeDataSet(Closeable ds) {
   public DiscoveryServiceClient getDiscoveryServiceClient() {
     return discoveryServiceClient;
   }
+
+  protected static Map<String, String> getMetricsContext(Program program, String runId) {
+    Map<String, String> tags = Maps.newHashMap();
+    tags.put(Constants.Metrics.Tag.APP, program.getApplicationId());
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: No tag for Namespace?

Comment (abaranau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: not yet

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractResourceReporter.java [hunk 3]
```diff
@@ -76,16 +89,15 @@ protected final ScheduledExecutorService executor() {
...
-  protected void sendMetrics(String context, int containers, int memory, int vcores, String runId) {
-    LOG.trace("Reporting resources in context {}: (containers, memory, vcores) = ({}, {}, {})",
-              context, containers, memory, vcores);
-    MetricsCollector collector = collectionService.getCollector(MetricsScope.SYSTEM, context, runId);
-    collector.gauge(METRIC_CONTAINERS, containers);
-    collector.gauge(METRIC_MEMORY_USAGE, memory);
-    collector.gauge(METRIC_VIRTUAL_CORE_USAGE, vcores);
+  protected void sendMetrics(Map<String, String> context, int containers, int memory, int vcores) {
+    LOG.trace("Reporting resources: (containers, memory, vcores) = ({}, {}, {})", containers, memory, vcores);
+    MetricsCollector metricsCollector = programMetricsCollectors.getUnchecked(context);
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractResourceReporter.java]: So there is no System metrics scope anymore? So can users not have a metrics name METRICS_CONTAINERS et al?

File: cdap-common/src/main/java/co/cask/cdap/common/hooks/MetricsReporterHook.java [hunk 4]
```diff
@@ -97,16 +100,21 @@ public void postCall(HttpRequest request, HttpResponseStatus status, HandlerInfo
         } else {
           name = "unknown";
         }
-        collector.increment("response." + name, 1, "status:" + code);
+
+        // todo: report metrics broken down by status
+        collector.increment("response." + name, 1/*, "status:" + code*/);
```
Comment (gokulavasan) [inline cdap-common/src/main/java/co/cask/cdap/common/hooks/MetricsReporterHook.java]: Was the inline commenting of code intentional?

**Comment by gokulavasan** [comment]
Minor comments. @gsps1 has one unaddressed comment. Otherwise LGTM :+1:

File: cdap-common/src/main/java/co/cask/cdap/common/hooks/MetricsReporterHook.java (hunk 4 above)
Comment (abaranau) [inline cdap-common/src/main/java/co/cask/cdap/common/hooks/MetricsReporterHook.java]: yes! Hence todo

Comment (abaranau) [inline cdap-common/src/main/java/co/cask/cdap/common/hooks/MetricsReporterHook.java]: will create jira for it

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceMetricsWriter.java [hunk 5]
```diff
@@ -190,15 +185,7 @@ private void reportIncrementSystemStats(Iterable<Counter> counters, MetricsColle
       // "<metric>" or "<metric>,<tag>" if tag is present
       String[] parts = counter.getName().split(",", 2);
       String metric = parts[0];
-      if (parts.length == 2) {
-        // has tag
-        String tag = parts[1];
-        collector.increment(metric, emitValue, tag);
-        int tagCountSoFar = (metricTagValues.containsKey(metric)) ? metricTagValues.get(metric) : 0;
-        metricTagValues.put(metric, tagCountSoFar + emitValue);
```
Comment (abaranau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceMetricsWriter.java]: I will address this along with the next PR.

**Pull request merged**